src/space_partition_exclude.c

```diff
@@ -0,0 +1,157 @@
+#include <postgres.h>
+#include <nodes/parsenodes.h>
+#include <nodes/nodeFuncs.h>
+#include <nodes/makefuncs.h>
+#include <parser/parsetree.h>
+#include <parser/parse_coerce.h>
+#include <parser/parse_func.h>
+#include <parser/parse_collate.h>
+#include <parser/parse_oper.h>
+#include <optimizer/clauses.h>
+#include <catalog/pg_type.h>
+#include <optimizer/paths.h>
+
+#include "cache.h"
+#include "hypertable.h"
+#include "partitioning.h"
+#include "compat.h"
+
+#include "space_partition_exclude.h"
+#include "chunk.h"
+#include "hypercube.h"
+#include "hypertable.h"
+
+typedef struct SpaceExcludeCtx
+{
+	Query	   *parse;
+	Hypertable *hentry;
+} SpaceExcludeCtx;
+
+
+
+static inline Dimension *
+get_dimension_for_partition_column_var(Var *var_expr, SpaceExcludeCtx *context)
+{
+	Hypertable *ht = context->hentry;
+	RangeTblEntry *rte = rt_fetch(var_expr->varno, context->parse->rtable);
+	char	   *varname;
+
+	if (rte->relid != ht->main_table_relid)
+		return NULL;
+
+	varname = get_rte_attribute_name(rte, var_expr->varattno);
+
+	return hyperspace_get_dimension_by_name(ht->space, DIMENSION_TYPE_CLOSED, varname);
+}
+
+SpacePartitionExclude *
+space_partition_exclude_get(RestrictInfo *base_restrict_info, Query *query, Hypertable *hentry)
+{
+	Expr	   *node = base_restrict_info->clause;
+
+	if (IsA(node, OpExpr))
+	{
+		OpExpr	   *exp = (OpExpr *) node;
+
+		if (list_length(exp->args) == 2)
+		{
+			/* only look at var op const or const op var; */
+			Node	   *left = (Node *) linitial(exp->args);
+			Node	   *right = (Node *) lsecond(exp->args);
+			Var		   *var_expr = NULL;
+			Node	   *other_expr = NULL;
+
+			if (IsA(left, Var))
+			{
+				var_expr = (Var *) left;
+				other_expr = right;
+			}
+			else if (IsA(right, Var))
+			{
+				var_expr = (Var *) right;
+				other_expr = left;
+			}
+
+			if (var_expr != NULL)
+			{
+				if (!IsA(other_expr, Const))
+				{
+					/* try to simplify the non-var expression */
+					other_expr = eval_const_expressions(NULL, other_expr);
+				}
+				if (IsA(other_expr, Const))
+				{
+					/* have a var and const, make sure the op is = */
+					Const	   *const_expr = (Const *) other_expr;
+					Oid			eq_oid = OpernameGetOprid(list_make2(makeString("pg_catalog"), makeString("=")), exprType(left), exprType(right));
+
+					if (eq_oid == exp->opno)
+					{
+						/*
+						 * I now have a var = const. Make sure var is a
+						 * partitioning column
+						 */
+						SpaceExcludeCtx context = (SpaceExcludeCtx) {
+							.parse = query,
```

Comment:
Is it safe to use pointers here (rather than copy), considering things like prepared statements, etc.? Just want to point out potential issue to think about.